search.go

```diff
@@ -32,11 +32,13 @@ func searchHandler(indexer Indexer, cacheTime time.Duration) func(w http.Respons
 			Filter: filter,
 		}
 
+		start := time.Now()
 		packages, err := indexer.Get(r.Context(), &opts)
 		if err != nil {
 			notFoundError(w, errors.Wrapf(err, "fetching package failed"))
 			return
 		}
+		util.SearchProcessDurationSeconds.Observe(time.Since(start).Seconds())
```

Comment:
Example of this metric:

```
# HELP epr_search_process_duration_seconds A histogram of package search processes.
# TYPE epr_search_process_duration_seconds histogram
epr_search_process_duration_seconds_bucket{le="0.005"} 23
epr_search_process_duration_seconds_bucket{le="0.01"} 23
epr_search_process_duration_seconds_bucket{le="0.025"} 23
epr_search_process_duration_seconds_bucket{le="0.05"} 23
epr_search_process_duration_seconds_bucket{le="0.1"} 23
epr_search_process_duration_seconds_bucket{le="0.25"} 23
epr_search_process_duration_seconds_bucket{le="0.5"} 23
epr_search_process_duration_seconds_bucket{le="1"} 23
epr_search_process_duration_seconds_bucket{le="2.5"} 23
epr_search_process_duration_seconds_bucket{le="5"} 23
epr_search_process_duration_seconds_bucket{le="10"} 23
epr_search_process_duration_seconds_bucket{le="+Inf"} 23
epr_search_process_duration_seconds_sum 0.012933142
epr_search_process_duration_seconds_count 23
```

Reply:
Let me challenge this.

Keep in mind that there are customers who need to deploy EPR and Prometheus `/metrics` is a standard endpoint that is easy to integrate with. APM isn't and we shouldn't require to install it additionally.

Reply:
Ok, I won't insist on adding the metric if it doesn't provide much value at the moment and there are other means to measure `/search` duration.

Regarding, the integration with APM - personally, I consider it as a heavy application to run and another component to maintaining, so I wouldn't install it. In the case of Elastic-hosted EPR, the APM is already present, so I wouldn't complain about it loud :)

Reply:
As @jsoriano points out, metric `epr_http_request_duration_seconds` is quite similar. The difference was mainly that this metric (`epr_search_process_duration_seconds`) was focused on the time to perform just the `indexer.Get()` call.

From that http request duration time, there are some processes that are not considered:
- create the filter from the query
- build the final array of packages (getPackageOutput)
- write headers
- print/write the response (data from the indexer)

I see that there is FilterPackages span, that is calculated inside the Apply() method of Filters. That would be something similar to what it is considered here.

Do you think that considering that this metric could lead to a worthy measurement?

Probably the metric name could be rephrased to be more meaningful, something like:
- epr_storage_get_duration_seconds
- epr_storage_indexer_get_duration_seconds

Following that, I could also add also a new metric to get an histogram about the update index process: `epr_storage_indexer_update_index_duration_seconds`.

Reply:
I've updated the name and narrow the scope of this metric.
The change can be seen here: https://github.com/elastic/package-registry/pull/827/commits/00315215f6f90bcd159d82fcd22baa53f0b6f7f5

As a proposal I've changed the name from "epr_search_process_duration_seconds" to "epr_storage_indexer_get_duration_seconds"

And this is the difference between these two metrics after running just this curl:
- curl https://localhost:8080/search

In this way, we can track all the get calls performed. There is one call more in epr_storage_indexer_get_duration_seconds because of the ensurePackageAvailable method.

```
# HELP epr_storage_indexer_get_duration_seconds A histogram of package search processes.
# TYPE epr_storage_indexer_get_duration_seconds histogram
epr_storage_indexer_get_duration_seconds_bucket{le="0.005"} 2
epr_storage_indexer_get_duration_seconds_bucket{le="0.01"} 2
epr_storage_indexer_get_duration_seconds_bucket{le="0.025"} 2
epr_storage_indexer_get_duration_seconds_bucket{le="0.05"} 2
epr_storage_indexer_get_duration_seconds_bucket{le="0.1"} 2
epr_storage_indexer_get_duration_seconds_bucket{le="0.25"} 2
epr_storage_indexer_get_duration_seconds_bucket{le="0.5"} 2
epr_storage_indexer_get_duration_seconds_bucket{le="1"} 2
epr_storage_indexer_get_duration_seconds_bucket{le="2.5"} 2
epr_storage_indexer_get_duration_seconds_bucket{le="5"} 2
epr_storage_indexer_get_duration_seconds_bucket{le="10"} 2
epr_storage_indexer_get_duration_seconds_bucket{le="+Inf"} 2
epr_storage_indexer_get_duration_seconds_sum 4.85e-07
epr_storage_indexer_get_duration_seconds_count 2

# HELP epr_http_request_duration_seconds A histogram of latencies for requests to the http server.
# TYPE epr_http_request_duration_seconds histogram
epr_http_request_duration_seconds_bucket{code="200",method="get",path="/search",le="0.005"} 0
epr_http_request_duration_seconds_bucket{code="200",method="get",path="/search",le="0.01"} 1
epr_http_request_duration_seconds_bucket{code="200",method="get",path="/search",le="0.025"} 1
epr_http_request_duration_seconds_bucket{code="200",method="get",path="/search",le="0.05"} 1
epr_http_request_duration_seconds_bucket{code="200",method="get",path="/search",le="0.1"} 1
epr_http_request_duration_seconds_bucket{code="200",method="get",path="/search",le="0.25"} 1
epr_http_request_duration_seconds_bucket{code="200",method="get",path="/search",le="0.5"} 1
epr_http_request_duration_seconds_bucket{code="200",method="get",path="/search",le="1"} 1
epr_http_request_duration_seconds_bucket{code="200",method="get",path="/search",le="2.5"} 1
epr_http_request_duration_seconds_bucket{code="200",method="get",path="/search",le="5"} 1
epr_http_request_duration_seconds_bucket{code="200",method="get",path="/search",le="10"} 1
epr_http_request_duration_seconds_bucket{code="200",method="get",path="/search",le="+Inf"} 1
epr_http_request_duration_seconds_sum{code="200",method="get",path="/search"} 0.008124299
epr_http_request_duration_seconds_count{code="200",method="get",path="/search"} 1
```

Reply:
An extra challenge/follow-up: if you can cover with metrics failed attempts of updating index/cursor, then the entire flow would be covered.